src/consensus/upgrades.h

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef ZCASH_CONSENSUS_UPGRADES_H
+#define ZCASH_CONSENSUS_UPGRADES_H
+
+#include "consensus/params.h"
+
+enum UpgradeState {
+    UPGRADE_DISABLED,
+    UPGRADE_PENDING,
+    UPGRADE_ACTIVE
+};
+
+struct NUInfo {
+    /** Branch ID (a random 32-bit value) */
+    uint32_t nBranchId;
+    /** User-facing name for the upgrade */
+    std::string strName;
+    /** User-facing information string about the upgrade */
+    std::string strInfo;
+};
+
+extern const struct NUInfo NetworkUpgradeInfo[];
+
+/**
+ * Checks the state of a given network upgrade based on block height.
+ * Caller must check that the height is >= 0 (and handle unknown heights).
+ */
+UpgradeState NetworkUpgradeState(
+    int nHeight,
```

Comment:
isn't an unsigned int better if it's supposed to be positive?

Reply:
Height is represented as an `int` throughout the codebase, and there are places where a negative height has a specific (non-height) meaning. Keeping this as an `int` ensures we can correctly assert, and that the caller can't accidentally pass in a negative value that gets coerced.